flips/context-propagation.md

```diff
@@ -0,0 +1,208 @@
+
+## Proposal
+
+### Separate lifecycle management from state observation
+
+Moving forward, we should split the `ReadyDoneAware` interface into two separate interfaces:
+- A `Runnable` interface which can be used to start and stop a component:
+  ```golang
+  type Runnable interface {
+    // Run will start the component
+    Run(context.Context)
+  }
+  ```
+- A `ReadyDoneAware` interface with methods `Ready` and `Done` which can be used to check a component's state (i.e wait for startup / shutdown to complete)
+  ```golang
+  type ReadyDoneAware interface {
+    // Ready returns a channel that will close when component startup has completed.
+    Ready() <-chan struct{}
+    // Done returns a channel that will close when component shutdown has completed.
+    Done() <-chan struct{}
+  }
+  ```
+
+A component will now be started by passing a `Context` to its `Run` method, and can be stopped by cancelling the `Context`. If a component needs to startup subcomponents, it can create child `Context`s from this `Context` and pass those to the subcomponents.
+#### Motivations
+- `Context`s are the standard way of doing go-routine lifecycle management in Go, and adhering to standards helps eliminate confusion and ambiguity for anyone interacting with the `flow-go` codebase. This is especially true now that we are beginning to provide API's and interfaces for third parties to interact with the codebase (e.g DPS).
+  - Even to someone unfamiliar with our codebase (but familiar with Go idioms), it is clear how a method signature like `Run(context.Context)` will behave. A method signature like `Ready()` is not so clear.
+  - If context propagation is done properly, there is no need to worry about any cleanup code in the `Done` method.
+- This allows us to separate the capability to check a component's state from the capability to start / stop it. We may want to give multiple other components the capability to check its state, without giving them the capability to start or stop it. Here is an [example](https://github.com/onflow/flow-go/blob/b50f0ffe054103a82e4aa9e0c9e4610c2cbf2cc9/engine/common/splitter/network/network.go#L112) of where this would be useful.
+- This provides a clearer way of defining ownership of components, and hence may potentially eliminate the need to deal with concurrency-safety altogether. Whoever creates a component should be responsible for starting it, and therefore they should be the only one with access to its `Runnable` interface. If each component only has a single parent that is capable of starting it, then we should never run into concurrency issues.
+- This enables us to define components which startup upon creation, and don't need to be explicitly `Run()` at a later point in time. Such components would only implement the `ReadyDoneAware` interface.
```

Comment:
I probably miss something, but if component starts up immediately and has no startup/shutdown code, why even bother with `ReadyDoneAware`? It can just be simple Go object without special lifecycle right?
We use `.Module()` in `NodeBuilder`, which is also badly named, as this is just a callback, nothing modular there.

Reply:
So this would refer to a component which *does* have startup / shutdown code, but the startup code isn't explicitly triggered by a call to `Run` (or `Start`, which is probably a better name as you suggested).

So instead of this:
```golang
ctx, cancel := context.WithCancel(context.Background())

// create component
component := NewComponent()

// start component
component.Start(ctx)

// wait for it to be ready
<-component.Ready()

// stop component
cancel()

// wait for it to be done
<-component.Done()
```

We *could* have something like this

```golang
ctx, cancel := context.WithCancel(context.Background())

// create and start component
component := NewComponent(ctx)

// wait for it to be ready
<-component.Ready()

// stop component
cancel()

// wait for it to be done
<-component.Done()
```